File: Anagrams/areTheseAnagramsHashTables.kt
fun main() {
    print("Enter string number 1: ")
    val stringInput1: String = readLine()!!
    println("Got it. Thanks!")
    print("Enter string number 2: ")
    val stringInput2: String = readLine()!!
    println(areTheseAnagramsHashTables(stringInput1, stringInput2))
}

fun areTheseAnagramsHashTables(s1: String, s2: String): String {
    val freq1 = mutableMapOf<Char, Int>()
    val freq2 = mutableMapOf<Char, Int>()
    if (s1.length != s2.length) {
        return ("These are not anagrams")
    }
    else {
        for (chA: Char in s1) {
            if (chA in freq1) {
                freq1.merge(chA, 1, Int::plus)
            }
            else {
                freq1[chA] = 1
            }
        }
        println(freq1.toList().sortedBy { (key, _) -> key }.toMap())
        for (chB: Char in s2) {
            if (chB in freq2) {
                freq2.merge(chB, 1, Int::plus)
            }
            else {
                freq2[chB] = 1
            }
        }
        println(freq2.toList().sortedBy { (key, _) -> key }.toMap())
        for (key: Char in freq1.keys) {
            if (key !in freq2.keys || freq1[key] != freq2[key]) {
                return "These are not anagrams"
            }
        }
        return "These are anagrams"
    }
}

File: Anagrams/areTheseAnagramsHashTables.go
package main

import (
	"fmt"
	"golang.org/x/exp/maps"
)

func main() {
	fmt.Println("Enter string number 1: ")
	var stringInput1 string
	_, _ = fmt.Scan(&stringInput1)
	fmt.Println("Got it. Thanks!")
	fmt.Println("Enter string number 2: ")
	var stringInput2 string
	_, _ = fmt.Scan(&stringInput2)
	fmt.Println(areTheseAnagramsHashTables(stringInput1, stringInput2))
}

func areTheseAnagramsHashTables(s1, s2 string) string {
	var freq1 = map[string]int{}
	var freq2 = map[string]int{}
	if len(s1) != len(s2) {
		return "These are not anagrams"
	} else {
		for _, chr := range s1 {
			if _, found := freq1[string(chr)]; found {
				freq1[string(chr)] += 1
			} else {
				freq1[string(chr)] = 1
			}
		}
		fmt.Println(freq1)
		for _, chr := range s2 {
			if _, found := freq2[string(chr)]; found {
				freq2[string(chr)] += 1
			} else {
				freq2[string(chr)] = 1
			}
		}
		fmt.Println(freq2)
		for _, chr := range maps.Keys(freq1) {
			if _, found := freq2[chr]; found == false {
				return "These are not anagrams"
			} else if freq1[chr] != freq2[chr] {
				return "These are not anagrams"
			}
		}
		return "These are anagrams"
	}
}
